systems/plants/@RigidBodyManipulator/contactConstraintDerivatives.m

```diff
@@ -1,13 +1,25 @@
-function [n, D, dn, dD] = contactConstraintDerivatives(obj, kinsol, idxA, idxB, xA, xB, normal, d, use_mex)
- compute_second_derivative = nargout > 2;
- 
+function [d, n, D, dn, dD] = contactConstraintDerivatives(use_mex, obj, normal, kinsol, idxA, idxB, xA, xB)
+ compute_second_derivative = nargout > 3;
+ tangents_only = (nargout == 1);
+
   if(use_mex)
+    if(tangents_only) 
+      d = contactConstraintsmex(normal);
+      return;
+    end
+
     if(~compute_second_derivative)
-      [n, D] = contactConstraintsmex(obj.mex_model_ptr, int32(idxA), int32(idxB), xA, xB, normal, d);
+      [d, n, D] = contactConstraintsmex(normal, obj.mex_model_ptr, int32(idxA), int32(idxB), xA, xB);
     else
-      [n, D, dn, dD] = contactConstraintsmex(obj.mex_model_ptr, int32(idxA), int32(idxB), xA, xB, normal, d);
+      [d, n, D, dn, dD] = contactConstraintsmex(normal, obj.mex_model_ptr, int32(idxA), int32(idxB), xA, xB);
     end
   else
+  d = obj.surfaceTangents(normal);
+
+  if(tangents_only) 
```

Comment:
I don't think there is currently an important case where we call the surfacetangents except from within contact constraints.  So i would just call it directly from within the contact constraints c++ and just let others call the matlab version

- Russ


> On Feb 22, 2015, at 3:04 PM, John Carter <notifications@github.com> wrote:
> 
> In systems/plants/@RigidBodyManipulator/contactConstraintDerivatives.m:
> 
> >      end
> >    else
> > +  d = obj.surfaceTangents(normal);
> > +
> > +  if(tangents_only) 
> This did come out messier than I expected. Surfacetangents depends only on the normals, so passing in a bunch of extra arguments in cases where it's the only thing calculated seems clunky. On the other hand, leaving it in complicates the input argument checking.
> 
> I'm starting to feel inclined to just leave them as 2 mex files for clarity.
> 
> —
> Reply to this email directly or view it on GitHub.
>